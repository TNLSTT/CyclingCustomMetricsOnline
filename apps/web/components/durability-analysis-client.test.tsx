import React, { type ReactNode } from 'react';
import { renderToString } from 'react-dom/server';
import { describe, expect, it, vi } from 'vitest';

import type { DurabilityAnalysisResponse } from '../types/durability-analysis';
import { DurabilityAnalysisClient } from './durability-analysis-client';

vi.mock('next-auth/react', () => ({
  useSession: () => ({ data: { accessToken: 'token-1' }, status: 'authenticated' }),
}));

const mockData: DurabilityAnalysisResponse = {
  ftpWatts: 265,
  filters: {
    minDurationSec: 10800,
    startDate: undefined,
    endDate: undefined,
    discipline: undefined,
    keyword: undefined,
  },
  disciplines: ['outdoor', 'trainer'],
  rides: [
    {
      activityId: 'ride-1',
      startTime: '2024-03-01T10:00:00Z',
      source: 'outdoor',
      durationSec: 14400,
      ftpWatts: 265,
      normalizedPowerWatts: 250,
      normalizedPowerPctFtp: 94.3,
      averagePowerWatts: 210,
      averageHeartRateBpm: 142,
      totalKj: 750,
      tss: 210,
      heartRateDriftPct: 3.5,
      bestLateTwentyMinWatts: 280,
      bestLateTwentyMinPctFtp: 105.7,
      durabilityScore: 92,
      segments: {
        early: {
          label: 'early',
          startSec: 0,
          endSec: 4320,
          durationSec: 4320,
          normalizedPowerWatts: 255,
          normalizedPowerPctFtp: 96.2,
          averagePowerWatts: 220,
          averageHeartRateBpm: 138,
          heartRatePowerRatio: 0.627,
        },
        middle: {
          label: 'middle',
          startSec: 4320,
          endSec: 10080,
          durationSec: 5760,
          normalizedPowerWatts: 248,
          normalizedPowerPctFtp: 93.6,
          averagePowerWatts: 208,
          averageHeartRateBpm: 140,
          heartRatePowerRatio: 0.673,
        },
        late: {
          label: 'late',
          startSec: 10080,
          endSec: 14400,
          durationSec: 4320,
          normalizedPowerWatts: 245,
          normalizedPowerPctFtp: 92.5,
          averagePowerWatts: 205,
          averageHeartRateBpm: 146,
          heartRatePowerRatio: 0.712,
        },
      },
      timeSeries: Array.from({ length: 6 }).map((_, index) => ({
        t: index * 900,
        power: 200 + index * 6,
        heartRate: 132 + index,
      })),
    },
  ],
};

vi.mock('swr', () => ({
  __esModule: true,
  default: vi.fn(() => ({ data: mockData, isLoading: false, error: undefined })),
}));

vi.mock('recharts', () => ({
  ResponsiveContainer: ({ children }: { children: ReactNode }) => <div>{children}</div>,
  AreaChart: ({ children }: { children: ReactNode }) => <div>{children}</div>,
  Area: () => <div data-testid="area" />,
  CartesianGrid: () => <div data-testid="grid" />,
  XAxis: () => <div data-testid="x-axis" />,
  YAxis: () => <div data-testid="y-axis" />,
  Tooltip: () => <div data-testid="tooltip" />,
  ReferenceArea: () => <div data-testid="reference" />,
  Line: () => <div data-testid="line" />,
  BarChart: ({ children }: { children: ReactNode }) => <div>{children}</div>,
  Bar: () => <div data-testid="bar" />,
  ScatterChart: ({ children }: { children: ReactNode }) => <div>{children}</div>,
  Scatter: () => <div data-testid="scatter" />,
  LineChart: ({ children }: { children: ReactNode }) => <div>{children}</div>,
  Legend: () => <div data-testid="legend" />,
}));

describe('DurabilityAnalysisClient', () => {
  it('renders a durability analysis summary snapshot', () => {
    const html = renderToString(
      <DurabilityAnalysisClient initialData={mockData} defaultMinDurationMinutes={180} />,
    );

    expect(html).toContain('Candidate rides');
    expect(html).toContain('Durability score formula');
    expect(html).toMatchSnapshot();
  });
});
