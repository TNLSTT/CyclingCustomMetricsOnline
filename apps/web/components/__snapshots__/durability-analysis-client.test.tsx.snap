// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DurabilityAnalysisClient > renders a durability analysis summary snapshot 1`] = `"<div class="space-y-6"><div class="rounded-lg border bg-card text-card-foreground shadow-sm"><div class="flex flex-col p-6 space-y-1"><h3 class="tracking-tight text-base font-semibold">Filter rides</h3><p class="text-sm text-muted-foreground">Focus on long rides and specific sessions to compare durability.</p></div><div class="p-6 pt-0 space-y-4"><div class="space-y-2"><div class="flex items-center justify-between text-sm font-medium text-foreground"><span>Minimum duration: <!-- -->180<!-- --> min</span><span class="text-muted-foreground">3.0<!-- --> hours</span></div><input type="range" min="60" max="420" step="15" class="w-full accent-primary" aria-label="Minimum ride duration in minutes" value="180"/></div><div class="grid gap-4 md:grid-cols-2"><label class="space-y-1 text-sm font-medium text-foreground"><span>Start date</span><input type="date" class="flex h-10 w-full rounded-md border border-input bg-transparent px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50" value=""/></label><label class="space-y-1 text-sm font-medium text-foreground"><span>End date</span><input type="date" class="flex h-10 w-full rounded-md border border-input bg-transparent px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50" value=""/></label></div><div class="grid gap-4 md:grid-cols-2"><label class="space-y-1 text-sm font-medium text-foreground"><span>Discipline / source</span><div class="flex items-center gap-2"><input class="flex h-10 w-full rounded-md border border-input bg-transparent px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50" list="durability-disciplines" placeholder="e.g. outdoor, trainer" value=""/><datalist id="durability-disciplines"><option value="outdoor"></option><option value="trainer"></option></datalist></div></label><label class="space-y-1 text-sm font-medium text-foreground"><span>Keyword search</span><input class="flex h-10 w-full rounded-md border border-input bg-transparent px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50" placeholder="Match ride source or metric key" value=""/></label></div><div class="flex justify-end"><button class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 ring-offset-background bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2" type="button">Apply filters</button></div></div></div><div class="rounded-lg border bg-card text-card-foreground shadow-sm"><div class="flex flex-col p-6 space-y-1"><h3 class="tracking-tight text-base font-semibold">Candidate rides</h3><p class="text-sm text-muted-foreground">Showing 1 rides that match the filters.</p></div><div class="p-6 pt-0"><div class="w-full overflow-auto"><table class="w-full caption-bottom text-sm"><thead class="[&amp;_tr]:border-b"><tr class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"><th class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&amp;:has([role=checkbox])]:pr-0 cursor-pointer" aria-sort="desc">Date</th><th class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&amp;:has([role=checkbox])]:pr-0">Source</th><th class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&amp;:has([role=checkbox])]:pr-0 cursor-pointer" aria-sort="none">Duration</th><th class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&amp;:has([role=checkbox])]:pr-0">NP (W)</th><th class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&amp;:has([role=checkbox])]:pr-0">Avg HR</th><th class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&amp;:has([role=checkbox])]:pr-0 cursor-pointer" aria-sort="none">Durability score</th></tr></thead><tbody class="[&amp;_tr:last-child]:border-0"><tr class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted bg-muted/70 cursor-pointer" data-state="selected"><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0"><div class="flex flex-col"><span class="font-medium text-foreground">Mar 1, 2024</span><span class="text-xs text-muted-foreground">10:00 AM</span></div></td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">outdoor</td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">4h 00m 00s</td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">250</td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">142</td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0"><div class="inline-flex items-center rounded-full border px-2.5 py-0.5 text-xs transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-primary text-primary-foreground shadow font-semibold">92</div></td></tr></tbody><caption class="mt-4 text-sm text-muted-foreground">Click a ride to view detailed durability diagnostics and charts.</caption></table></div></div></div><div class="space-y-6"><div class="rounded-lg border bg-card text-card-foreground shadow-sm"><div class="p-6 space-y-4 pt-6"><div class="flex flex-wrap items-center justify-between gap-4"><div><h2 class="text-2xl font-semibold text-foreground">Selected ride</h2><p class="text-sm text-muted-foreground">Mar 1, 2024<!-- --> · <!-- -->10:00 AM<!-- --> · Source: <!-- -->outdoor</p></div><div class="flex items-center gap-3 text-sm"><div class="text-right"><p class="text-muted-foreground">Durability score</p><div class="inline-flex items-center rounded-full border px-2.5 py-0.5 text-xs transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-primary text-primary-foreground shadow font-semibold">92</div></div><div class="text-right"><p class="text-muted-foreground">Late 20-min power</p><p class="font-semibold text-foreground">105.7<!-- -->% FTP</p></div><div class="text-right"><p class="text-muted-foreground">HR drift</p><p class="font-semibold text-foreground">3.5%</p></div></div></div><div class="grid gap-4 md:grid-cols-4 text-sm"><div><p class="text-muted-foreground">Duration</p><p class="font-semibold text-foreground">4h 00m 00s</p></div><div><p class="text-muted-foreground">TSS (est)</p><p class="font-semibold text-foreground">210</p></div><div><p class="text-muted-foreground">Total energy</p><p class="font-semibold text-foreground">750.0<!-- --> kJ</p></div><div><p class="text-muted-foreground">Normalized power</p><p class="font-semibold text-foreground">250<!-- --> W (<!-- -->94.3<!-- -->% FTP)</p></div></div></div></div><div class="grid gap-6 lg:grid-cols-2"><div class="rounded-lg border bg-card text-card-foreground shadow-sm lg:col-span-2"><div class="flex flex-col space-y-1.5 p-6"><h3 class="tracking-tight text-base font-semibold">Power &amp; heart rate over time</h3><p class="text-sm text-muted-foreground">Highlighted regions show the opening and closing thirds used for durability scoring.</p></div><div class="p-6 pt-0 h-[320px]"><div><div><defs><linearGradient id="powerGradient" x1="0" x2="0" y1="0" y2="1"><stop offset="5%" stop-color="var(--primary)" stop-opacity="0.25"></stop><stop offset="95%" stop-color="var(--primary)" stop-opacity="0"></stop></linearGradient></defs><div data-testid="grid"></div><div data-testid="x-axis"></div><div data-testid="y-axis"></div><div data-testid="y-axis"></div><div data-testid="tooltip"></div><div data-testid="reference"></div><div data-testid="reference"></div><div data-testid="area"></div><div data-testid="line"></div></div></div></div></div><div class="rounded-lg border bg-card text-card-foreground shadow-sm"><div class="flex flex-col space-y-1.5 p-6"><h3 class="tracking-tight text-base font-semibold">Early vs. late normalized power</h3><p class="text-sm text-muted-foreground">Compare FTP-relative output across the opening and closing thirds.</p></div><div class="p-6 pt-0 h-[280px]"><div><div><div data-testid="grid"></div><div data-testid="x-axis"></div><div data-testid="y-axis"></div><div data-testid="tooltip"></div><div data-testid="bar"></div></div></div></div></div><div class="rounded-lg border bg-card text-card-foreground shadow-sm"><div class="flex flex-col space-y-1.5 p-6"><h3 class="tracking-tight text-base font-semibold">Segment metrics</h3><p class="text-sm text-muted-foreground">Detailed breakdown across the early, middle, and late thirds.</p></div><div class="p-6 pt-0"><div class="w-full overflow-auto"><table class="w-full caption-bottom text-sm"><thead class="[&amp;_tr]:border-b"><tr class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"><th class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&amp;:has([role=checkbox])]:pr-0">Segment</th><th class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&amp;:has([role=checkbox])]:pr-0">NP (% FTP)</th><th class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&amp;:has([role=checkbox])]:pr-0">Avg power (W)</th><th class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&amp;:has([role=checkbox])]:pr-0">Avg HR (bpm)</th><th class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&amp;:has([role=checkbox])]:pr-0">HR:Power</th></tr></thead><tbody class="[&amp;_tr:last-child]:border-0"><tr class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 capitalize">early</td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">96.2</td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">220</td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">138</td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">0.627</td></tr><tr class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 capitalize">middle</td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">93.6</td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">208</td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">140</td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">0.673</td></tr><tr class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0 capitalize">late</td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">92.5</td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">205</td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">146</td><td class="p-4 align-middle [&amp;:has([role=checkbox])]:pr-0">0.712</td></tr></tbody></table></div></div></div></div></div><div class="grid gap-6 lg:grid-cols-2"><div class="rounded-lg border bg-card text-card-foreground shadow-sm"><div class="flex flex-col space-y-1.5 p-6"><h3 class="tracking-tight text-base font-semibold">Duration vs. durability</h3><p class="text-sm text-muted-foreground">Each point represents a filtered ride.</p></div><div class="p-6 pt-0 h-[280px]"><div><div><div data-testid="grid"></div><div data-testid="x-axis"></div><div data-testid="y-axis"></div><div data-testid="tooltip"></div><div data-testid="scatter"></div></div></div></div></div><div class="rounded-lg border bg-card text-card-foreground shadow-sm"><div class="flex flex-col space-y-1.5 p-6"><h3 class="tracking-tight text-base font-semibold">Durability trend over time</h3><p class="text-sm text-muted-foreground">Track whether late-ride resilience is improving.</p></div><div class="p-6 pt-0 h-[280px]"><div><div><div data-testid="grid"></div><div data-testid="x-axis"></div><div data-testid="y-axis"></div><div data-testid="tooltip"></div><div data-testid="legend"></div><div data-testid="line"></div></div></div></div></div></div><div class="rounded-lg border bg-card text-card-foreground shadow-sm"><div class="flex flex-col space-y-1.5 p-6"><h3 class="tracking-tight text-base font-semibold">Durability score formula</h3><p class="text-sm text-muted-foreground">How we translate the metrics above into a single 0–100 score.</p></div><div class="p-6 pt-0 text-sm text-muted-foreground space-y-2"><p>Scores start at 100 points. We subtract half a point for each percentage point drop in normalized power (as % FTP) from the opening to closing thirds, and we deduct 0.75 points for every percentage point of positive heart-rate drift. To recognize strong finishes, we add half a point for every percentage point that the best 20-minute power in the final third exceeds FTP. The score is clamped between 0 and 100.</p><p>These insights assume the rider maintained a steady pacing intent throughout the ride and leveraged the recorded FTP value. Large coasting sections or inaccurate FTP entries may skew the analysis.</p></div></div></div>"`;
